compiler/backend/src/org/jetbrains/kotlin/codegen/SamWrapperCodegen.java

```diff
@@ -133,7 +133,7 @@ public Type genWrapper(
         );
         cv.visitSource(file.getName(), null);
 
-        WriteAnnotationUtilKt.writeSyntheticClassMetadata(cv, state);
+        WriteAnnotationUtilKt.writeSyntheticClassMetadata(cv, state, isInsideInline);
```

Comment:
Shouldn't we also check here that it's inside a non-private inline function?

Reply:
Sure, I can add that. Alternatively, I was thinking that going forward we could only generate inline SAM wrappers from within public inline functions. Since these classes are regenerated at call sites this wouldn't be a breaking change either and might improve SAM wrapper caching. What do you think?

Reply:
So you mean changing the logic which computes `isInsideInline` in `SamWrapperClasses.getSamWrapperClass` to check only if it's inside public inline? Yes, sounds good too!

Reply:
Great, done! I've also added tests for this, and got rid of the odd behavior where we would generate an inline sam wrapper when it was used from a lambda argument to an inline function. I'm really confused as to why we did that.

Reply:
I somehow didn't run all tests for this change and missed that this broke `kt24825`. Incidentally, that test is probably the reason for the JVM BE behavior around SAM wrappers in inline lambdas.

Unresolving this thread until I can figure out what went wrong there.

Reply:
The bug was in the inliner, which identified non-inline SAM wrappers as anonymous objects to be regenerated and would then remove the original wrapper.

Since this only affects non-public inline functions and hence only inline calls from within the same model I just fixed this in the inliner.